1-intro/intermediate.ipynb

```diff
@@ -0,0 +1,354 @@
+   "metadata": {},
+   "source": [
+    "### Sets\n",
+    "Sets in Python work lite sets from set theory. That is, they represent an unordered collection of unique elements. This means that if you try to add an element to a set that is already present, the set will not store a duplicate; it will just keep one instance of that element. Sets do not maintain any specific order of their elements. Beware of this if you convert data to sets and back.\n",
+    "You can create a set using curly braces {} or the set() constructor. Elements can be added or removed using the `.add()` and `.remove()` methods respectively."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "2d307a6d-2ccd-4148-9c02-bb1d0c2e734b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "my_set = {1, 2, 3}\n",
+    "another_set = set([3, 4, 5])\n",
+    "print(my_set)\n",
+    "\n",
+    "my_set.add(4)\n",
+    "print(my_set)\n",
+    "\n",
+    "my_set.remove(2)\n",
+    "print(my_set)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e32203e8-1b3c-4b5d-a5aa-308a2eb54d2d",
+   "metadata": {},
+   "source": [
+    "Sets support various set operations, such as `union()`, `intersection()`, and `difference()`. These operations allow you to combine and compare sets in useful ways. For example:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "cb11f0af-6deb-4686-ace4-9955192bff28",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "my_set = {1, 2, 3}\n",
+    "\n",
+    "union_set = my_set.union(another_set)\n",
+    "print(union_set)\n",
+    "\n",
+    "intersection_set = my_set.intersection(another_set)\n",
+    "print(intersection_set)\n",
+    "\n",
+    "difference_set = my_set.difference(another_set)\n",
+    "print(difference_set)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f37af5fd-554e-4df2-9a23-a25ca4832a65",
+   "metadata": {},
+   "source": [
+    "### Lambda functions\n",
+    "Lambda functions are small, anonymous functions that are defined using the `lambda` keyword. They are often used for short, simple operations and can be created on the fly without needing a name. A lambda function can also be assigned to a variable in order to be reusable.\n",
+    "A lambda function has the following basic structure: `lambda arguments: expression`.\n",
+    " - arguments: These are the input parameters that the lambda function takes. You can have zero or more arguments, separated by commas.\n",
+    " - expression: This is the operation or calculation that the lambda function performs using the input arguments.\n",
+    "Lambda functions are typically used for small, single-line operations. For example, you can create a lambda function that adds two numbers:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "e5cca364-0576-4f2c-bb5e-93fb2040a705",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "add = lambda x, y: x + y\n",
```

Comment:
Assigning lambda functions to variables is a code smell, see [PEP 8](https://peps.python.org/pep-0008/#programming-recommendations) and [`Ruff rule E731`](https://docs.astral.sh/ruff/rules/lambda-assignment/).